### PR TITLE
Object storage updates

### PR DESCRIPTION
**Pull request by scross01**
- Adds support for including custom metadata headers on containers
- Adds support for including custom metadata headers on objects
- Adds support for Qutoas on containers
- Fixes creation of objects using "CopyFrom" option

### CODE REVIEW

File: storage/container.go [hunk 1]
```diff
@@ -65,17 +125,38 @@ func (c *StorageClient) CreateContainer(input *CreateContainerInput) (*Container
...
-	headers["X-Container-Meta-Temp-URL-Key"] = input.PrimaryKey
-	headers["X-Container-Meta-Temp-URL-Key-2"] = input.SecondaryKey
-	headers["X-Container-Meta-Access-Control-Allow-Origin"] = strings.Join(input.AllowedOrigins, " ")
-	headers["X-Container-Meta-Access-Control-Expose-Headers"] = strings.Join(input.ExposedHeaders, " ")
-	headers["X-Container-Meta-Access-Control-Max-Age"] = strconv.Itoa(input.MaxAge)
+	headers[hTempURLKey] = input.PrimaryKey
+	headers[hTempURLKey2] = input.SecondaryKey
+	headers[hAccessControlAllowOrigin] = strings.Join(input.AllowedOrigins, " ")
+	headers[hAccessControlExposeHeaders] = strings.Join(input.ExposedHeaders, " ")
+	// headers[hPolicyGeoreplication] = strings.Join(input.GeoreplicationPolicy, " ")
```
Comment (mbfrahry) [inline storage/container.go]: Is there a reason this is commented out?

File: storage/container.go [hunk 2]
```diff
@@ -156,17 +268,35 @@ func (c *StorageClient) UpdateContainer(input *UpdateContainerInput) (*Container
...
-	headers["X-Container-Meta-Access-Control-Expose-Headers"] = strings.Join(input.ExposedHeaders, " ")
-	headers["X-Container-Meta-Access-Control-Max-Age"] = strconv.Itoa(input.MaxAge)
+	c.updateOrRemoveStringValue(headers, hTempURLKey, input.PrimaryKey)
+	c.updateOrRemoveStringValue(headers, hTempURLKey2, input.SecondaryKey)
+	c.updateOrRemoveStringValue(headers, hAccessControlAllowOrigin, strings.Join(input.AllowedOrigins, " "))
+	c.updateOrRemoveStringValue(headers, hAccessControlExposeHeaders, strings.Join(input.ExposedHeaders, " "))
+	c.updateOrRemoveIntValue(headers, hAccessControlMaxAge, input.MaxAge)
+	c.updateOrRemoveIntValue(headers, hQuotaBytes, input.QuotaBytes)
+	c.updateOrRemoveIntValue(headers, hQuotaCount, input.QuotaCount)
+	// c.updateOrRemove(headers, hPolicyGeoreplication, strings.Join(input.GeoreplicationPolicy, " "))
```
Comment (mbfrahry) [inline storage/container.go]: Same here

File: storage/container_test.go [hunk 3]
```diff
@@ -118,8 +146,17 @@ func TestAccContainerLifeCycle(t *testing.T) {
...
+		t.Fatalf(fmt.Sprintf("Max Age do not match Wanted: %d Recieved: %d", updatedMaxAge, container.MaxAge))
+	}
+	if container.QuotaBytes != updatedQuotaBytes {
+		t.Fatalf(fmt.Sprintf("Quota Bytes do not match Wanted: %d Recieved: %d", updatedQuotaBytes, container.QuotaBytes))
+	}
+	if container.QuotaCount != updatedQuotaCount {
+		t.Fatalf(fmt.Sprintf("Quota Count do not match Wanted: %d Recieved: %d", updatedQuotaCount, container.QuotaCount))
+	}
+	if diff := pretty.Compare(container.CustomMetadata, updatedCustomMetaData); diff != "" {
+		t.Fatalf(fmt.Sprintf("CustomMetadata diff (-got +want)\n%s", diff))
```
Comment (mbfrahry) [inline storage/container_test.go]: Got an error from the test here 
```
	container_test.go:159: CustomMetadata diff (-got +want)
		 {
		  Abc-Def: "123",
		  Bar: "foo",
		- Foo: "bar",
		 }
```

File: storage/container.go (hunk 1 above)
Comment (scross01) [reply storage/container.go]: The Georeplication Policy header set on the container, but not covered in the API docs, so commented it out for now till its clear how this attribute should be used.

Comment (mbfrahry) [reply storage/container.go]: Perfect

File: storage/container_test.go (hunk 3 above)
Comment (scross01) [reply storage/container_test.go]: Fixed. Update to support explicit removal of custom metadata headers, so `Foo` now gets removed.

**Comment by scross01** [comment]
UpdateContainer() now accepts a list of custom metadata headers to be explicitly removed.

**Comment by mbfrahry** [comment]
Hmmm, I'm not sure I like having the user pass in what to delete. What're your thoughts on hitting the api to see what is being removed and then sending that to the remove custom data header?

**Comment by scross01** [comment]
Passing the the explicit remove header requests maps better to the API. So I think it's cleaner from the API/SDK perspective.  From the provider calling the SDK its easy to determine which headers to remove based on the state diff.

**Pull request merged**